From 3e77299435152113f66bb079f4e103422a2751a9 Mon Sep 17 00:00:00 2001
From: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>
Date: Fri, 22 Dec 2023 11:52:25 -0800
Subject: [PATCH] wifi: mac80211: Fix station dump fields in WDS mode

After WDS AP gets associated with WDS STA, DTIM and beacon interval
values in the station dump is seen as 0.

During the WDS Association, DTIM and beacon interval stats are being referred
from the VLAN interface instead of the AP's interface. Fix this issue by
referring to the AP's interface to get the DTIM and beacon interval values.

Signed-off-by: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>

diff --git a/net/mac80211/sta_info.c b/net/mac80211/sta_info.c
index e41eb45..0c5f9a1 100644
--- a/net/mac80211/sta_info.c
+++ b/net/mac80211/sta_info.c
@@ -2758,8 +2758,16 @@ void sta_set_sinfo(struct sta_info *sta, struct station_info *sinfo,
 
 		sinfo->filled |= link_sta_filled;
 	} else {
-		dtim_period = sdata->deflink.conf->dtim_period;
-		beacon_interval = sdata->deflink.conf->beacon_int;
+		if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN) {
+			struct ieee80211_sub_if_data *ap_sdata;
+
+			ap_sdata = container_of(sdata->bss, struct ieee80211_sub_if_data, u.ap);
+			dtim_period = ap_sdata->deflink.conf->dtim_period;
+			beacon_interval = ap_sdata->deflink.conf->beacon_int;
+		} else {
+			dtim_period = sdata->deflink.conf->dtim_period;
+			beacon_interval = sdata->deflink.conf->beacon_int;
+		}
 		sinfo->filled |= link_sta_set_info(&sta->deflink, sinfo, true);
 	}
 
-- 
2.25.1

